course-content/21_getting_data_with_pandas.qmd

```diff
@@ -0,0 +1,427 @@
+
+Because of this, it is common to have a folder exclusively for raw data. Secondary folders may then be set-up for 'cleaned' data, python scripts, analysis outputs and more. 
+:::
+
+```{python}
+df = pd.read_csv('../data/gapfinder.csv')
+df
+```
+
+:::{.callout-tip}
+## .. ?
+
+The `..` used in the above path tells python to look in the directory directly above the one you are currently in - the one where your notebook is saved. 
+
+As an example, my directory structure looks like:
+```
+Intro to Python
+│
+└─── introduction_to_python
+│   │   hello_world.ipynb
+│   │   hello_world.py
+│   
+└─── introduction_to_pandas
+│   │   intro_to_pandas.ipynb  <------ (THIS FILE!)
+│
+└─── data
+│   │   gapfinder.csv <--------------- (The file we want access to!)
+```
+
+This file is in a different subfolder relative to the csv, so we first 'back out' of this folder using `..`, and then 'enter' the data folder using a regular file path. This is called **relative pathing** and can be useful for accessing data within a single project that will always be in the same spot! 
+:::
+
+This dataset has 6 columns, 1704 rows, and a mixture of different datatypes. Just like we were able to access the index of a series, we can do the same with a dataframe. Now, we can also access (and change if need be!) the column names as well:
+
+```{python}
+display(df.index)
+display(df.columns)
+
+df.columns = ['country', 'year', 'pop', 'continent', 'life expectancy', 'gdpPercap']
+display(df.columns)
+```
+
+We can access each of these series individually if we want. There are two ways to access a series in a dataframe - either with square bracket indexing, or treating the column name as an `attribute`:
+
+```{python}
+display(df['country'])
+display(df.country)
+```
+
+:::{.callout-tip}
+## No Spaces! 
+
+<center>
+    <img src="images/getting-data-with-pandas/no_capes.jpg" 
+    width=300 style="margin:auto"/>
+    <p style="text-align: center">
+    </p>
+</center>
+
+As a general rule of thumb, we never want to include special characters such as spaces, periods, hyphens and so on in our column names, as this will alter pandas capability of calling each of the columns as an attribute. Above, you will notice we reset one of the columns to have a space in the name. If we try to access this series as an attribute now, it will fail:
+
+```{python}
+df['life expectancy']
+```
+
+```{python}
+#| error: true
+df.life expectancy
+```
+
+Try to stick to lowercase letters, numbers and underscores for your column names! (And as a matter of fact, let us change it back now):
```

Comment:
maybe point out not to start names with numbers (although including them in the names is ok practice)